ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,572 @@
+        drcontext, sizeof(drstatecmp_user_data_t));
+    memset(data, 0, sizeof(*data));
+    *user_data = (void *)data;
+
+    /* Determine whether the basic block is free of side effects. */
+    bool side_effect_free = true;
+    for (instr_t *inst = instrlist_first_app(bb); inst != NULL;
+         inst = instr_get_next_app(inst)) {
+        side_effect_free &= !drstatecmp_may_have_side_effects_instr(inst);
+    }
+    /* Cannot easily execute twice the bb if it has side-effects. Thus, no need to
+     * duplicate.
+     */
+    if (!side_effect_free) {
+        data->side_effect_free = false;
+        return DR_EMIT_DEFAULT;
+    }
+    data->side_effect_free = true;
+
+    /* Duplication process.
+     * Consider the following example bb:
+     *   instr1
+     *   instr2
+     *   term_instr
+     *
+     * In this stage, we just duplicate the bb (except for its terminating
+     * instruction) and add special labels to the original and duplicated blocks. The
+     * duplicated bb is for now skipped with a jump. The insert instrumentation stage will
+     * remove this jump, and add saving/restoring of machine state and the state
+     * comparison.
+     * Note that there might be no term_instr (no control transfer instruction) and the bb
+     * just falls-through. Even with no term_instr the jmp and the TERM label are inserted
+     * in the same way, as shown in this example.
+     *
+     * The example bb is transformed, in this stage, as follows:
+     * ORIG_BB:
+     *   instr1
+     *   instr2
+     *   jmp TERM
+     *
+     * COPY_BB:
+     *   instr1
+     *   instr2
+     * TERM:
+     *   term_instr
+     *
+     */
+
+    /* Create a clone of bb. */
+    instrlist_t *copy_bb = instrlist_clone(drcontext, bb);
+
+    /* Create and insert the labels. */
+    drstatecmp_insert_labels(drcontext, bb, instrlist_first(bb), DRSTATECMP_LABEL_ORIG_BB,
+                             /*preinsert=*/true);
+    instr_t *copy_bb_label =
+        drstatecmp_insert_labels(drcontext, copy_bb, instrlist_first(copy_bb),
+                                 DRSTATECMP_LABEL_COPY_BB, /*preinsert=*/true);
+    instr_t *term_inst_copy_bb = instrlist_last_app(copy_bb);
+    instr_t *term_label = NULL;
+    if (instr_is_cti(term_inst_copy_bb) || instr_is_return(term_inst_copy_bb)) {
+        term_label = drstatecmp_insert_labels(drcontext, copy_bb, term_inst_copy_bb,
+                                              DRSTATECMP_LABEL_TERM, /*preinsert=*/true);
+    } else {
+        term_label = drstatecmp_insert_labels(drcontext, copy_bb, term_inst_copy_bb,
+                                              DRSTATECMP_LABEL_TERM, /*preinsert=*/false);
+    }
+    /* Need an operand for the jmp instr targeting this label. */
+    opnd_t term_label_opnd = opnd_create_instr(term_label);
+
+    /* Replace the terminating instruction of the original bb with a jmp to the
+     * terminating instruction of the bb clone (if any).
```

Comment:
What is the purpose of this added jump?  Other instrumentation libraries/clients are not likely to look at it and conclude that it skips the copy and then ignore the copy: they'll assume linear control flow and blindly instrument every instruction in the list, ignoring the fact that there's a jump here.

Reply:
@derekbruening Without the jmp, one can have tools, like drreg, which will consider the orig and copy blocks as one whole basic block and therefore perform incorrect analysis (for drreg's case liveness). Think of a situation where a register used in the orig block is erroneously considered dead only because the copy block writes to it further down the line.

Edit: I just read that you raised a similar point further down in your review.

Reply:
The jump is not used anymore.